pallets/staking/src/lib.rs

```diff
@@ -111,53 +174,157 @@ pub mod module {
             &self,
             storage: &mut sp_runtime::Storage,
         ) -> Result<(), String> {
-            <Self as frame_support::traits::GenesisBuild<T>>::assimilate_storage(self, storage)
+            <Self as GenesisBuild<T>>::assimilate_storage(self, storage)
         }
     }
 
-    #[pallet::pallet]
-    pub struct Pallet<T>(PhantomData<T>);
-
     #[pallet::hooks]
-    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
-        fn on_finalize(_now: T::BlockNumber) {}
-    }
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
+        /// Put assets under staking, the native assets will be transferred to the account
+        /// owned by the pallet, user receive voucher in return, such vocher can be further
+        /// used as collateral for lending.
+        ///
+        /// - `amount`: the amount of staking assets
         #[pallet::weight(10_000)]
         #[transactional]
         pub fn stake(origin: OriginFor<T>, amount: Balance) -> DispatchResultWithPostInfo {
-            let who = ensure_signed(origin)?;
-            Self::stake_internal(&who, amount)?;
+            let sender = ensure_signed(origin)?;
+
+            let exchange_rate = ExchangeRate::<T>::get();
+            let voucher_amount = exchange_rate
+                .reciprocal()
+                .and_then(|r| r.checked_mul_int(amount))
+                .ok_or(Error::<T>::InvalidExchangeRate)?;
+
+            T::Currency::transfer(
+                T::StakingCurrency::get(),
+                &sender,
+                &Self::account_id(),
+                amount,
+            )?;
+            T::Currency::deposit(T::LiquidCurrency::get(), &sender, voucher_amount)?;
+            TotalVoucher::<T>::try_mutate(|b| -> DispatchResult {
+                *b = b.checked_add(voucher_amount).ok_or(Error::<T>::Overflow)?;
+                Ok(())
+            })?;
+            TotalStakingAsset::<T>::try_mutate(|b| -> DispatchResult {
+                *b = b.checked_add(amount).ok_or(Error::<T>::Overflow)?;
+                Ok(())
+            })?;
 
+            Self::deposit_event(Event::Staked(sender, amount));
             Ok(().into())
         }
 
+        /// Withdraw assets from liquid staking pool for offchain relay chain nomination.
+        ///
+        /// May only be called from `T::WithdrawOrigin`.
+        ///
+        /// - `agent`: the multisig account of relay chain.
+        /// - `amount`: the requested assets.
         #[pallet::weight(10_000)]
         #[transactional]
-        pub fn unstake(origin: OriginFor<T>, amount: Balance) -> DispatchResultWithPostInfo {
-            let who = ensure_signed(origin)?;
-            Self::unstake_internal(&who, amount)?;
+        pub fn withdraw(
```

Comment:
it looks like that people can stake 1 dot then withdraw 100dot as there is no limit?

Reply:
This function can only be called by a specific collective, the collective may be multisig accounts in parachain, which also have a corresponding account in relay chain.